### ISSUE TITLE
Lazy segment tree is difficult to modify for other operations

### ISSUE THREAD

**Issue by IvanRenison** [kth-competitive-programming/kactl#227]
The lazy segment tree of this is repo very useful when exactly the operations that it has are needed, but when other are needed it is difficult to modify because it is not very clear where the lazy values are propagated and where the operations are made.
In normal segment tree on the other hand is very easy to modify the operation because it is in a separate function.
I propose making a similar separation for lazy segment tree.
Probably it wold look something like this: https://github.com/IvanRenison/icpc-team-notebook-el-vasito/blob/update_STree_lazy/data_structures/segment_tree_lazy.cpp (note that that lazy segment tree is different because it is not lazy in memory allocation)

If you are interested in this improvement I can make it (including the tests) and send a pull request.

**Comment by BarrensZeppelin** [comment]
> Does that :+1: mean that you are interested in this improvement?

Indeed. I've opened https://github.com/kth-competitive-programming/kactl/issues/216#issue-1444507915 which is similar.

**Comment by IvanRenison** [comment]
Great!
